### ISSUE TITLE
Research into uchan

### ISSUE THREAD

**Issue by DonaldTsang** [DangerOnTheRanger/maniwani#107]
https://github.com/Floens/uchan is a pre-existing Python based imageboard software
It uses PostgreSQL + TypeScript + Memcache + Varnish
Might be good to use it as a type of reference on what features are needed.

**Comment by DangerOnTheRanger** [comment]
Looking over the [roadmap](https://github.com/Floens/uchan/issues/1), I don't really see much that Maniwani either doesn't do already, or doesn't do but should (in my opinion). The plugin system is probably the biggest thing I wish Maniwani did, though uchan's plugin system is [pretty minimal](https://github.com/Floens/uchan/blob/dev/uchan/lib/plugin_manager.py). Sometimes that's all that's needed, though.

**Comment by DonaldTsang** [comment]
@DangerOnTheRanger what do you think should not be implemented?

**Comment by DangerOnTheRanger** [comment]
Hard reliance on memcached (no real reason not to use Redis instead today, in my opinion) is probably the biggest one. The example infrastructure is tied somewhat tightly to Varnish, which is fine, but less flexible than I'd like for Maniwani since a design goal I've had from the start is that throwing up an example deployment is as simple as `docker run dangerontheranger/maniwani` from an empty system, and requiring an external cache/database to be present would preclude that.

**Comment by samip5** [comment]
I think the plugin part of uchan should be implemented in Maniwani at least as well organize the repo better.

**Comment by DangerOnTheRanger** [comment]
Cleaning up the repo would be something worth doing, definitely. I'm busy/have been busy with #77, though I'm on the tail end of getting the `react-ui` branch to feature parity with the Jinja2/JQuery UI- so I might be able to tackle that soon, and maybe add some developer's documentation on Maniwani's architecture, that sort of thing.

As far as I know, uchan's plugin system imports Python modules but only provides around 2 places for plugins to hook into (post creation and general JS injection, I think). I'm open to having a plugin system, but I'd ideally like some good discussion on what sort of hooks Maniwani should provide, both for better brainstorming of what potential plugins could do, as well as the amount of work required to implement the plugin system itself.

**Comment by samip5** [comment]
So what happens if a client has no js enabled? I hope that the plan is to  support non-javascript users aswell even when react-ui goes live.

**Comment by DangerOnTheRanger** [comment]
Yeah, I don't want to leave Noscript users out in the cold, and right now the `react-ui` branch takes that into account by running an Express server (as a separate executable inside the Maniwani container when devmode is active or as an entirely separate container in production mode) that renders the React components server-side before sending it down to the client. If the client has JS enabled, then the UI will be dynamically updated client-side - things like relative post timestamps will go from "1 minute ago" to "5 minutes ago," and new posts/threads will be automatically displayed - but that's it. Since the React components themselves hit the client as fully-formed HTML, even without JS enabled they look like what you'd expect.

The goal is to keep the user experience discrepancy roughly the same as it is now - you can browse and use a Maniwani instance perfectly fine without needing JS, but get some extra real-time features if you turn it on.
